src/centralized_telescope.rs

```diff
@@ -0,0 +1,421 @@
+        for t in 0..setup.d {
+            if limit == setup.b {
+                return (limit, None);
+            }
+            if let Some(r) = Round::new(v, t, setup.n_p, setup.sec_param) {
+                let (l, proof_opt) = Self::dfs(setup, &bins, &r, limit.saturating_add(1));
+                if proof_opt.is_some() {
+                    return (l, proof_opt);
+                }
+                limit = l;
+            }
+        }
+        (limit, None)
+    }
+
+    /// Alba's proving algorithm, based on a depth-first search algorithm.
+    /// Calls up to setup.r times the prove_index function and returns an empty
+    /// proof if no suitable candidate is found.
+    pub fn prove(setup: &Setup, set: &[Element]) -> Option<Self> {
+        (0..setup.r).find_map(|v| Self::prove_index(setup, set, v).1)
+    }
+
+    /// Alba's verification algorithm, follows proving algorithm by running the
+    /// same depth-first search algorithm.
+    pub fn verify(setup: &Setup, proof: &Self) -> bool {
+        if proof.t >= setup.d || proof.v >= setup.r || proof.items.len() as u64 != setup.u {
+            return false;
+        }
+        if let Some(mut round) = Round::new(proof.v, proof.t, setup.n_p, setup.sec_param) {
+            for element in proof.items.clone() {
+                if let Some(h) = Self::h0(setup, proof.v, element) {
+                    if round.h_u64 == h {
+                        if let Some(r) = Round::update(&round, element, setup.sec_param) {
+                            round = r;
+                        } else {
+                            return false;
+                        }
+                    } else {
+                        return false;
+                    }
+                } else {
+                    return false;
+                }
+            }
+            return Self::h2(setup, &round);
+        }
+        false
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::gen_items;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::{RngCore, SeedableRng};
+
+    #[test]
+    fn test_verify() {
+        let mut rng = ChaCha20Rng::from_seed(Default::default());
+        let nb_tests = 1_000;
+        let set_size = 1_000;
+        for _t in 0..nb_tests {
+            let seed = rng.next_u32().to_be_bytes().to_vec();
+            let s_p = gen_items::<DATA_LENGTH>(&seed, set_size);
+            let params = Params {
+                lambda_sec: 10.0,
+                lambda_rel: 10.0,
+                n_p: 80,
+                n_f: 20,
+            };
+            let setup = Setup::new(&params);
+            let proof = Proof::prove(&setup, &s_p).unwrap();
+            assert!(Proof::verify(&setup, &proof.clone()));
+            let proof_d = Proof {
+                v: proof.v,
+                t: proof.t.wrapping_add(1),
+                items: proof.items.clone(),
+            };
+            assert!(!Proof::verify(&setup, &proof_d));
```

Comment:
Suggest putting each bad proof in its own scope. This also solves the problem with having to come up with new variable names.
```suggestion
            {
              let proof = Proof {
                  v: proof.v,
                  t: proof.t.wrapping_add(1),
                  items: proof.items.clone(),
              };
              assert!(!Proof::verify(&setup, &proof));
            }
```

Reply:
That does not bring much. Feel free to do this in the unit test issue I assigned to you: https://github.com/cardano-scaling/alba/issues/49